### ISSUE TITLE
Correctly displaying responsive release notes

### ISSUE THREAD

**Issue by ThomasBarnekow** [vslavik/winsparkle#98]
I'm trying to display release notes published on a WordPress site that uses a fully responsive design. The release notes are displayed correctly on an iPhone and in Internet Explorer 11 when I resize to the width of WinSparkle's release notes control.

When displayed by WinSparkle, however, the release notes are not rendered correctly. The browser control shows a horizontal scrollbar and displays only about 60% of the page width. And the page is a bit blurry.

This does not happen with the plain HTML release notes based on which I created the WordPress post. No horizontal scrollbar is shown.

I'd be thankful for any hints on how this could be fixed (e.g., by configuring the control in a different way).

**Comment by vslavik** [comment]
> I'd be thankful for any hints

Well, apply your skills. You're a developer, debugging things is what you *do* — so do. Nobody can offer you a magical solution to an ill-described (more like non-described) issue with a completely unknown input.

If you still believe (*after* doing meaningful investigation) that whatever this is is somehow WinSparkle's fault, file a proper, *reproducible* bug report — see http://www.chiark.greenend.org.uk/~sgtatham/bugs.html on how to do it (I would expect this to be second nature for a developer too, but apparently not, because instead of an URL or any details you write that it's "responsive design" as if that meant anything).

> I'd be thankful for any hints on how this could be fixed 

Most likely, simply fix your broken CSS. This is *extremely* unlikely to have any relation to WinSparkle. More likely than not, you simply have poor CSS code for MSIE.

**Comment by ThomasBarnekow** [comment]
@vslavik I'm not sure why you are always so defensive in your responses and why you like to offend people (I have read your responses in a number of issue reports). You are the most unfriendly person I have seen on GitHub or any other platform. I will therefore no longer contribute anything to WinSparkle.

**Comment by vslavik** [comment]
@ThomasBarnekow I’m not “always” (as you know full well because you actually submitted a small PR before) “defensive” and if truth “offends” you, that’s your issue, not mine, particularly after being rude yourself. I simply don’t particularly appreciate it when somebody wastes my time with poor or even bogus issues. Dealing with one stupid issue is no big deal, right? Well, multiply that by 10 years and many tickets and it adds up quickly. Have some empathy for the *other* side of it.

Forget about any feelings of entitlement and try to look at this issue objectively: Did you do a good job of describing the problem? No — it’s impossible to reproduce because you withheld essential information. Should a *developer* know better than that? Absolutely yes (end users don’t and can’t be expected to, but end users don’t use WinSparkle). Did you follow the best practice of first blaming your own code, making sure that’s not the case, then reporting *your findings* (including reasons to believe a bug elsewhere)? You did not.

Instead, you decided it’s a problem with the webview control and decided it may be configurable somehow to avoid it, without getting into any particulars  — even though a quick search of the code would tell you it’s a `IWebBrowser2` and any issues with it can’t be WinSparkle specific. After learning that, you could then search the existing body of knowledge about the control to see if anything could explain what you’re seeing. And *then* you would be able to file an issue, or a PR, describing what *specifically* is wrong with the interaction of the control with your CSS *and here’s what needs to be done to fix it*. Or you would learn about some issue with the bits of CSS that you have (and nobody else knows!) with MSIE and fixed that instead.

In any case, you would be considerate (i.e., polite) towards the maintainers of the project that is useful to you. And even if you *didn’t* find any clue, it would be immediately obvious from the issue that you made the effort to diagnose it and couldn’t, that you made your homework, and that there really is something that needs to be looked into. And you would have provided a test case to check.

So: please do and reopen the issue when you do. That’s more productive than getting offended over being (rightly) called out about your issue’s deficiencies.

**Comment by joshkel** [comment]
@vslavik - Unfortunately,  I can kind of understand why @ThomasBarnekow said what he did. :sweat:

I do understand where you're coming from. I'm impressed with what you've accomplished with WinSparkle (and Poedit). I'm sure it can be draining and frustrating to feel like you're expected to give free support for products that you've already released for free. I probably don't appreciate _enough_ just how frustrating it can be, because my own open source efforts have been more niche and less sustained. I'm a little envious at your accomplishments. :smile:

But several of your responses on WinSpark issues just come across as... rude, for lack of a better term. Some of the replies don't just say that you'd need more information to diagnose the problem (that would obviously be fine). They don't just say that you don't have time to do the work being asked and giving the other person a pointer on how they can proceed (that would be understandable). Instead, they make it sound like you're offended that the person even dared to ask the question in the form given, and it sounds like you question the developer's credentials as a whole, as if it's a personal failing on their part. Maybe that's not the intent, but that's how it sounds to me as an observer. Then you close the issue, which comes across as being uninterested in further discussion or in offering any more assistance (at least, in my perception, as someone who's not  extremely familiar with GitHub etiquette).

Saying that a bug report doesn't have enough information to reproduce, or that you're not interested in working on a feature request yourself but would welcome a PR, or whatever, is fine.  But saying that developers are being stupid, lazy, idiots, entitled, etc. in the issues they raise isn't constructive.  There's a good chance it's not true (because, let's face it, anyone can miscommunicate or forget a detail or have a bad day or still be learning one particular aspect of our wide and wonderful field). Even if it is true, it puts people on the defensive instead of encouraging meaningful change.

There may also be some differences of expectations going on. If I drew an imaginary spectrum from "I will do your homework for you" to "Go away and stop bothering me," you're probably further to the right than a lot of open source developers. Absolutely nothing wrong with that! Like you said, people have no right to feel entitled simply because they got something for free. But some issues are likely posted, not out of a sense of entitlement, but simply out of assuming developers are elsewhere on the spectrum; those probably deserve a polite explanation.

To put it another way - I totally understand not wanting to have time taken up by issues that aren't very high quality.  But a form letter explaining that your time is limited and politely instructing the submitter to investigate/reproduce/submit a PR themselves first would take even less time, and it would probably be more productive in the long run.

Just my 2 cents.

**Comment by joshkel** [comment]
@ThomasBarnekow - A couple of wild guesses: It sounds like it might be a DPI issue (see #88), and I know that IE may put itself into compatibility mode when used as an embedded web browser (see [here](http://stackoverflow.com/q/6914664/25507)). Depending on what all changes in compatibility mode, that might cause problems for responsive sites.

**Comment by vslavik** [comment]
> They don't just say that you don't have time to do the work being asked 

It’s not a matter of “having time”, it’s a matter of being expected to do free work *that the submitter could and should have done on their own*. Or to list a few entitlements you probably don’t even realize you said: it’s unreasonable to expect me to do any work, or to give pointers to googleable stuff, or to be "interested in discussion” or to actively offer assistance. These takes a lot of *time* and when you come off as taking them for granted, as the kids these days (including you, seeing how you casually drop all those in) do, it *very* strongly discourages from doing any Open Source stuff. I can’t recommend [reading this entertaining rant](https://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/) enough. In short: I publish code as a good Open Source citizen. In exchange, I expect users to behave as good Open Source citizens too. Is that so much to ask?

> Then you close the issue, which comes across as being uninterested in further discussion or in offering any more assistance

And issue is an *issue* — something *actionable*. When it’s clear (or appears to be) that the actual issue is outside of WinSparkle, all I can do with such issue is to close it to get it off my TODO list. That’s normal, doesn’t have any connotations you project into it, and is reversible with the helpful *Reopen* button.

> anyone can miscommunicate or forget a detail

There's a difference between forgetting a detail (which is fine and happens, as you say) and not even putting a basic report together. As I said, I don’t expect normal users to know how to do. I very much do expect a *developer* to, because they *know* how to do it properly. And when they skip the homework, it’s very, very obvious they are telling me that they don’t value my time nearly as much as their own (despite saving so much time by using my project). That, to quote you, comes across as rude to *me*.

> But a form letter explaining that your time is limited and politely instructing the submitter to investigate/reproduce/submit a PR themselves first would take even less time

*Now* that GitHub finally has support for this, yes, it’s worth spending an hour crafting that. That’s a *very* new development, though, and it wasn’t available until very recently.

**Comment by vslavik** [comment]
> A couple of wild guesses: It sounds like it might be a DPI issue (see #88), 

Nope. That could cause either tiny text or blurry, but correctly scaled, but not being too large.

> and I know that IE may put itself into compatibility mode when used as an embedded web browser

It pretty much always does, in one form or another, and this is well known — the embedded browser’s mode is controlled in the same way MSIE is, via `X-UA-Compatible`, hence my assumption this is neither WinSparkle nor `IWebBrowser2` issue, but a MSIE compatibility one.

**Comment by ThomasBarnekow** [comment]
@joshkel Thanks for commenting. Your 2 cents are very much appreciated.

**Comment by ThomasBarnekow** [comment]
> And issue is an issue — something actionable. When it’s clear (or appears to be) that the actual issue is outside of WinSparkle, all I can do with such issue is to close it to get it off my TODO list. That’s normal, doesn’t have any connotations you project into it, and is reversible with the helpful Reopen button.

@vslavik While I did not provide exhaustive detail in my initial description of the issue, I said that the web page is correctly displayed in certain browsers, e.g., Safari on iOS and Internet Explorer 11 (MSIE), when those are resized to the width of WinSparkle's embedded browser (or to an even smaller width in the case of the iPhone). And, by the way, Safari on Mac OS X works as well (e.g., when resizing to the smallest width possible). Of all the browsers I've tested, WinSparkle's embedded browser is the only one that does not display the page correctly. Would you say that this permits the hypothesis that there might be an issue with WinSparkle or one of its components?

> I'd be thankful for any hints on how this could be fixed (e.g., by configuring the control in a different way).

>> Most likely, simply fix your broken CSS. This is extremely unlikely to have any relation to WinSparkle. More likely than not, you simply have poor CSS code for MSIE.

You can have a look at the [Release Notes](http://contractarchitect.com/category/release-notes/) I am trying to display. You will see that they are displayed perfectly well in MSIE and other browsers (which is what I said in my initial description). The CSS is neither "poor" nor "broken". The CSS is not even "mine", because it is produced by the [Enterprise Pro Theme](http://my.studiopress.com/themes/enterprise/), i.e., a child theme of the [Genesis Framework](http://www.studiopress.com/) used to create responsive [WordPress](http://wordpress.org/)-based sites.

I'm not sure how you want to proceed with this. I can't reopen the issue because I can't see a _Reopen_ button.

**Comment by vslavik** [comment]
> I'm not sure how you want to proceed with this. 

I’m not particularly eager to spend any time helping you after what transpired here. But oh well, let’s try again, even though I’ll probably regret it:

> Would you say that this permits the hypothesis that there might be an issue with WinSparkle or one of its components?

Not necessarily, considering the easily obtainable knowledge that WinSparkle uses an OS browser component (and so any rendering issues are beyond its control) and well-known/well-googleable `WebBrowser` use of older MSIE engine by default...

> You can have a look at the Release Notes I am trying to display.

Please observe how I asked for a test case (i.e. *in* WinSparkle) and only got a random link instead. Please also observe how we suggested likely MSIE engine issues, offered a pointer to a likely fix (`X-UA-Compatible`), yet you completely ignored that (as evidenced by the provided link, which doesn’t use it). I don’t understand why you ask for "any hints on how this could be fixed” and then ignore them?

So I did what you should have done before your latest comment, I forced `WebBrowser` to render in MSIE 11 mode (I didn’t know how, by the way, but googled it in a few minutes) and guess what: as suspected, that changed the rendering and removed the scrollbar.

> The CSS is neither "poor" nor "broken". The CSS is not even “mine"

The CSS on *your* server, under *your* control, serving *your* release notes is “yours” and the W3C validator [begs to differ](http://www.css-validator.org/validator?uri=http%3A%2F%2Fcontractarchitect.com%2Fcategory%2Frelease-notes%2F&profile=css3&usermedium=all&warning=1&lang=en) with your bold claim about its quality: 33 errors (when treated as CSS3; 97 otherwise) is broken and 2557 warnings is poor quality, unquestionably. By the way: if I were debugging rendering of *my* release notes, this would be the first thing I would check...

> I can't reopen the issue because I can't see a Reopen button.

Oh, must be owner-only then, but just as well: the above investigation shows the issue is indeed invalid. 

You have a couple of options:

1. Fix your CSS to be valid. May be enough.
2. Use `X-UA-Compatible`. Should work.
3. Force `WebBrowser` to use different engine for your executable.
4. Or do the sensible thing and use minimal CSS for release notes. Your users won’t appreciate having most of the window covered with some header and having to scroll to read any actual notes anyway.

**Comment by vslavik** [comment]
> that "Starting with IE11, document modes are considered deprecated and should no longer be used." Thus, that was not really an option.

You seem to be refusing to accept the fact that *WebBrowser control is not IE11* despite repeatedly being told so. Please do accept it already.

> How did you force the WebBrowser control to render in IE11 mode?

[Seriously?](http://lmgtfy.com/?q=webbrowser+engine+force)

**Comment by ThomasBarnekow** [comment]
> You seem to be refusing to accept the fact that WebBrowser control is not IE11 despite repeatedly being told so. Please do accept it already.

I understand and accept that. But please also understand and accept that I don't want to design my web site for WinSparkle. If possible, I would like WinSparkle to display a specific page (the release notes) of a web site that is rendered totally fine (in all widths) by modern browsers such as Safari (on Mac OS and iOS), Edge, or IE11.

> Seriously?

You could have just shared what you found (or did). But you chose the fun way. Nice.

So, referring back to my original post, the `WebBrowser` control was indeed configured in a different way.  In its default Internet Explorer 7 (IE7, launched in October 2006) rendering mode, the `WebBrowser` control used by WinSparkle does not fully support HTML5 (released in October 2014) or the more recent CSS releases. When configured to use IE11 rendering mode (IE9 works as well), the `WebBrowser` control correctly displays the web page.

So, we do have an issue with "WinSparkle or one of its components" (i.e., the `WebBrowser` control). Fortunately, we also have a fix (i.e., a registry setting used to configure the `WebBrowser` control).

It would make sense to document that fix. I could do that, if you like.

**Comment by vslavik** [comment]
> But please also understand and accept that I don't want to design my web site for WinSparkle

That's what web developers (i.e. people doing non-trivial styling) *do*: they work with the user agents they have. This is the control *virtually everybody* uses for embedding on Windows.

> So, we do have an issue with "WinSparkle or one of its components"... It would make sense to document that fix.

No, the *freaking operating system* is not a "WinSparkle component". Stop spinning for excuses. There's no "fix" and this is not the place to document standard Windows behavior, and forcing the entire host process' behavior is not a good idea (unlike, you know, the *other* fixes, the ones you ignored).

> You could have just shared what you found (or did). But you chose the fun way. Nice.

OK. Enough is enough. I *did* point you in the right direction, you ignored it, because why would you bother if you can ask somebody else to digest it for you, right? You expect everything to be served to you on a silver plate. You didn't even *try*! I'm not your servant and I won't stand for being treated like one. This is *extremely* disrespectful and offending from you.

Enough time was wasted. I'm locking this issue.
